### ISSUE TITLE
Classpath override

### ISSUE THREAD

**Issue by christopinka** [michaelruocco/property-loader#2]
Nicely done. unless I'm missing it, would be good to have override capability like

`Properties p = new Properties();
InputStream in = this.getClass().getResourceAsStream("c.properties");
p.load(in);

String externalFileName = System.getProperty("app.properties");
InputStream fin = new FileInputStream(new File(externalFileName));
p.load(fin);`

Your program would be invoked similar to this:

java -jar app.jar -Dapp.properties="/path/to/custom/app.properties"

**Comment by christopinka** [comment]
Move Fakes to test package?

**Comment by christopinka** [comment]
Wow. Build is nice in gradle. Can I ask what the inspiration was for the project?

**Comment by christopinka** [comment]
I get it now. I just override by calling it twice. Closing. Oops. Anyway. Nice job.

**Comment by michaelruocco** [comment]
With the library as it stands, you could implement the code you have above by using the code in the property package here: https://github.com/michaelruocco/file-loader/tree/master/src/main/java/uk/co/mruoc/file/property and doing the following:

```
Properties p = new Properties();

PropertyLoader classpathLoader = new ClasspathPropertyLoader();
p.putAll(classpathLoader.load("c.properties"))

PropertyLoader fileSystemLoader = new FileSystemPropertyLoader();
p.putAll(fileSystemLoader.load(System.getProperty("app.properties")));
```

There is also a static helper to have having to create instances of the loader, but for some reason I hadn't added the properties methods to it. So I have raised [this PR](https://github.com/michaelruocco/file-loader/pull/3) which would allow the above to be shortened to:

```
Properties p = new Properties();
p.putAll(FileLoader.loadPropertiesFromClasspath("c.properties"));
p.putAll(FileLoader.loadPropertiesFromFileSystem(System.getProperty("app.properties")));
```

It also moves the fakes into the test fixtures package too, and renames and moves the top level file loader class. So I would intend to increment the major version number when merging this. Is this the sort of thing you had in mind?

**Comment by christopinka** [comment]
Yes it is. Thanks. What are your thoughts on handling nullpointerexception? When no system property "app.properties" I have a stacktrace that looks like

```
Exception in thread "main" java.lang.NullPointerException
	at sun.misc.MetaIndex.mayContain(MetaIndex.java:242)
	at sun.misc.URLClassPath$JarLoader.getResource(URLClassPath.java:1058)
	at sun.misc.URLClassPath.getResource(URLClassPath.java:250)
	at sun.misc.URLClassPath.getResource(URLClassPath.java:303)
	at java.lang.ClassLoader.getBootstrapResource(ClassLoader.java:1257)
	at java.lang.ClassLoader.getResource(ClassLoader.java:1086)
	at java.lang.ClassLoader.getResource(ClassLoader.java:1084)
	at java.net.URLClassLoader.getResourceAsStream(URLClassLoader.java:233)
	at uk.co.mruoc.file.ClasspathInputStreamLoader.load(ClasspathInputStreamLoader.java:14)
	at uk.co.mruoc.file.property.DefaultPropertyLoader.load(DefaultPropertyLoader.java:19)
	at uk.co.mruoc.file.property.ClasspathPropertyLoader.load(ClasspathPropertyLoader.java:17)
```

Could do a null check but might be nice if null then return empty.

**Comment by michaelruocco** [comment]
The library would expect a path to be passed to load the properties, if you don't have a path configured I would handle that in your calling code personally. It depends on whether code expects that property to be set or not.

If you code expects the property to be set then I would handle it by throwing an error e.g.

```
       private static void main(String[] args) {
            Properties p = new Properties();
            p.putAll(FileLoader.loadPropertiesFromClasspath("c.properties"));
            p.putAll(FileLoader.loadPropertiesFromFileSystem(getAppPropertiesPath()));
       }
               
       private static String getAppPropertiesPath() {
           return Optional.ofNullable(System.getProperty("app.properties"))
               .orElseThrow(() -> new IllegalStateException("app.properties must be configured"));
       }
```

Of if you just want to load the other properties and ignore app.properties if not set, probably using some private methods to break up the complexity, so something like this:

```
       private static void main(String[] args) {
            Properties p = new Properties();
            p.putAll(FileLoader.loadPropertiesFromClasspath("c.properties"));
            getAppProperties().ifPresent(appProperties -> p.putAll(appProperties));
       }     
         
       private static Optional<Properties> getAppProperties() {
           return getAppPropertiesPath().map(path -> FileLoader.loadPropertiesFromFileSystem(path));
       }
                
       private static Optional<String> getAppPropertiesPath() {
           return Optional.ofNullable(System.getProperty("app.properties"));
       }
```

I wouldn't want to update the existing behaviour directly to return empty if not present as that might be misleading to other clients. In the case of the path not being supplied at all that seems like an edge case that should be handled in the client code.

**Comment by christopinka** [comment]
Possibly

```
 @Override
    public Properties load(String path) {
        try (InputStream stream = inputStreamLoader.load(path)) {
            return converter.toProperties(stream);
        } catch (IOException | UncheckedIOException e) {
            throw new PropertiesNotFoundException(path, e);
        } catch (NullPointerException e) {
            throw new PropertiesNotFoundException("file path to load is null", e);
        }


    }
```

**Comment by christopinka** [comment]
What's your opinion on making exceptions part of the api?

**Comment by christopinka** [comment]
oh. these are unchecked. ok. makes sense. does it?

Just thinking that if app.properties isn't set will be null path. I suppose null check makes the most sense.

**Comment by michaelruocco** [comment]
Its quite a big debate, but generally checked exceptions are not a good idea in my opinion, rather than explaining here, here are a few articles that discuss it in more detail:

https://medium.com/@shley_ng/clean-code-error-handling-19fec47ad688#:~:text=Checked%20Exceptions&text=If%20you%20throw%20a%20checked,changes%20on%20many%20higher%20levels.
https://ahmadatwi.me/2017/02/05/how-to-handle-java-exceptions-in-clean-code-part-2/

If it was going to handle nulls I would do a check up front rather than catching null pointer exception e.g.

```
    @Override
    public Properties load(String path) {
        if (StringUtils.isEmpty(path)) {
            throw new IllegalArgumentException("path should not be null or empty");
        }
        try (InputStream stream = inputStreamLoader.load(path)) {
            return converter.toProperties(stream);
        } catch (IOException | UncheckedIOException e) {
            throw new PropertiesNotFoundException(path, e);
        }
    }
```

However, looking at the stack trace that gets produced when null is passed in, I would say that is reasonably obvious already that it is caused by the path being null, but it could be updated to do the above instead. Feel free to raise a PR to add this check if you wish? If not Ill see about adding it when I get some time.

**Comment by michaelruocco** [comment]
Actually, don't worry about the PR, I noticed a few small issues with the existing error handling, so I have cleaned that up and have added the null / empty string validation check at the same time. Ill push the updates shortly.

**Comment by michaelruocco** [comment]
Changes pushed and will be available in maven central shortly on version 7.0.2
